### PR TITLE
Make wxUIActionSimulator works more reliably on GTK/X11

### PR DESCRIPTION
**Pull request by AliKet**

### CODE REVIEW

File: tests/controls/gridtest.cpp [hunk 1]
```diff
@@ -1187,6 +1181,9 @@ TEST_CASE_METHOD(GridTestCase, "Grid::ResizeScrolledHeader", "[grid]")
 
     wxYield();
     sim.MouseMove(point);
+#ifdef __WXGTK20__
+    sim.MouseMove(point);
+#endif
```
Comment (vadz) [inline tests/controls/gridtest.cpp:1186]: This is really weird, but I guess you don't know why is this needed neither?

File: tests/controls/gridtest.cpp [hunk 2]
```diff
@@ -145,38 +145,12 @@ TEST_CASE_METHOD(GridTestCase, "Grid::CellEdit", "[grid]")
...
-        if ( sw.Time() > 250 )
-        {
-            WARN("Editor control not shown until timeout expiration");
-            break;
-        }
-
-        wxYield();
-    }
-#endif // __WXGTK__
+    wxYield();
```
Comment (vadz) [inline tests/controls/gridtest.cpp:148]: It's really nice that this hack is not needed any more!

File: src/unix/uiactionx11.cpp [hunk 3]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+        : m_display(display), m_isMotion(false)
+    {
+        static int s_maxDepressed = 0;
+
+        if ( depressed )
+        {
+            ++ms_numDepressed; // A key or button is down.
+
+            if ( ms_numDepressed - s_maxDepressed != 1 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:79]:
```suggestion
            wxASSERT_MSG( ms_numDepressed - s_maxDepressed == 1,
                                      "Invalid call to wxXSync() ctor");
```

I think moving it above and writing the condition as just `ms_numDepressed == s_maxDepressed` would be even more clear though.

File: src/unix/uiactionx11.cpp [hunk 4]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+public:
+    wxXSync(wxX11Display& display)
+        : m_display(display), m_isMotion(true)
+    {
+    }
+
+    wxXSync(wxX11Display& display, bool depressed)
+        : m_display(display), m_isMotion(false)
+    {
+        static int s_maxDepressed = 0;
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:72]: I don't understand why is this needed, is it just for debugging wrong use of this class (and what does wrong actually mean here anyhow? It would be nice to document this in the class comment above...)?

File: src/unix/uiactionx11.cpp [hunk 5]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+        {
+            delay = 0;
+        }
+
+        return delay;
+    }
+
+private:
+    wxX11Display& m_display;
+    const bool    m_isMotion;
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:149]: I'd like to add a comment explaining that if it is `false`, it's button or key event because it isn't immediately obvious right now.

File: src/unix/uiactionx11.cpp [hunk 6]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+        }
+
+        wxYield();
+
+        // Default_Delay might not be sufficient for some
+        // tests which need more additionnal time to pass.
+        wxMilliSleep(Default_Delay + GetAdditionalDelay());
+    }
+
+    static long GetAdditionalDelay()
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:129]: Passing extra delay from one part of the program to another via environment variables is rather strange. Why not just have some public `static void SetAdditionalDelay(long delay)` instead and call it from the testing code?

Also, why should we set _additional_ delay instead of just setting the delay? I feel like this is more complicated than necessary and we should just have some `ms_delay`, initialized with `Default_Delay`, and then functions for getting/setting it.

File: src/unix/uiactionx11.cpp [hunk 7]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+    {
+        XSync(m_display, False);
+
+        m_isMotion ? OnMotion() : OnKeyOrButtonUp();
+    }
+
+private:
+    static void OnMotion()
+    {
+        wxMilliSleep(Motion_Delay);
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:110]: Is it intentional that this doesn't use "additional delay"?

File: src/unix/uiactionx11.cpp [hunk 8]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+            if ( ms_numDepressed < 0 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
+        }
+    }
+
+    ~wxXSync()
+    {
+        XSync(m_display, False);
+
+        m_isMotion ? OnMotion() : OnKeyOrButtonUp();
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:104]: I find this code slightly confusing, why do we need these `OnXXX()` methods? This naming convention is usually reserved for the event handlers, but they don't handle anything... I'd call them `DoAfterXXX()` or something like that.

But, in fact, looking at their code, I'd probably just inline them here.

File: src/unix/uiactionx11.cpp [hunk 9]
```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
...
+    }
+
+private:
+    static void OnMotion()
+    {
+        wxMilliSleep(Motion_Delay);
+        wxYield();
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:111]: It's interesting that for the motion events you sleep first and then yield, while for the other ones it's the reverse. Do you have any idea about why does it need to be done like this? If so, it would be great to have a comment explaining it here.

**Review by vadz — COMMENTED** [review]
Thanks a lot for working on this!

If the tests pass, I'm ready to merge this in its current state, without even addressing all the comments below (although I'd probably try to fix at least some of them myself), because this is just too useful to have.

I admit I still don't understand what was the actual problem which prevented things from working with GTK 3 completely before, is it just `wxSetInputFocusToXWindow()` addition that fixed it or something else? But, again, this is just for my own curiosity, what matters is that the tests pass and if they do, it's really great news.

Thanks again!

**Comment by AliKet** [comment]
> I admit I still don't understand what was the actual problem which prevented things from working with GTK 3 completely before, is it just `wxSetInputFocusToXWindow()` addition that fixed it or something else?

Yes `wxSetInputFocusToXWindow ()` fixes what appears to be broken on Xvfb display.

In fact, `xdpyinfo` was my tool for spotting the discrepancy between the Xvfb and the real X display, making `XTest` hardly not working on an Xvfb display:

```
xdpyinfo -display :0 | less

name of display:    :0
version number:    11.0
vendor string:    The X.Org Foundation
vendor release number:    11906000
X.Org version: 1.19.6
⋮
keycode range:    minimum 8, maximum 255
focus:  window 0x2e03eb1, revert to Parent
number of extensions:    29
    BIG-REQUESTS
    Composite
:
```
whereas on Xvfb i have this output:
```
xdpyinfo -display :99 | less

name of display:    :99
version number:    11.0
vendor string:    The X.Org Foundation
vendor release number:    11906000
X.Org version: 1.19.6
⋮
keycode range:    minimum 8, maximum 255
focus:  PointerRoot
number of extensions:    25
    BIG-REQUESTS
    Composite
    DAMAGE
:
```
Notice how the focus is set to `PointerRoot` for the `Xvfb`, and according to documentation:

> If focus is PointerRoot, the focus window is dynamically taken to be the root window of whatever screen the pointer is on at each keyboard event. In this case, the revert_to argument is ignored.

which explains why some key presses and button clicks are lost for certain Xvfb setups and not for others.

File: tests/controls/bitmaptogglebuttontest.cpp [hunk 10]
```diff
@@ -88,7 +88,9 @@ void BitmapToggleButtonTestCase::Click()
     CPPUNIT_ASSERT(m_button->GetValue());
 
     clicked.Clear();
+#ifndef __WXGTK__
     wxMilliSleep(1000);
+#endif
```
[the remark this answers is not in the document]
Comment (AliKet) [reply tests/controls/bitmaptogglebuttontest.cpp:93]: I'll remove this.

File: tests/controls/gridtest.cpp (hunk 1 above)
Comment (AliKet) [reply tests/controls/gridtest.cpp:1186]: I did some tracing and found that, for a given window, gtk_window_motion_notify_callback is not called at all due to a fake motion unless the pointer is already on the same window. and unless i am missing something, without motion event, wxGrid won't be able to begin a drag operation. this is the reason for the second call to MouseMove(). sadly, even if it works now locally, i failed to make it work on Travis CI !

File: src/unix/uiactionx11.cpp (hunk 4 above)
Comment (AliKet) [reply src/unix/uiactionx11.cpp:72]: `s_maxDepressed` helps to :

- detect erroneous calls to `KeyUp/MouseUp` before a prior calls to `KeyDown/MouseDown`.
- make sure all keys are down before trying to release any key.
- sleep only once: when all keys are down and after the last depressed key is up.

For ex.: 
`sim.Char('F', wxMOD_ALT);
`treated as
```
KeyDown('F');
KeyDown(wxMOD_ALT);
wxYield();
wxMilliSleep(Default_Delay);
KeyUp(wxMOD_ALT);
KeyUp('F');
wxYield();
wxMilliSleep(Default_Delay);
```

File: src/unix/uiactionx11.cpp (hunk 7 above)
Comment (AliKet) [reply src/unix/uiactionx11.cpp:110]: Yes this is intentional.

File: src/unix/uiactionx11.cpp (hunk 9 above)
Comment (AliKet) [reply src/unix/uiactionx11.cpp:111]: My fault, this is unintentionally reversed.
but now I am wondering what would happen if we just `wxYield`ing without any delay at all ?
I'll take a closer look.

Comment (AliKet) [reply src/unix/uiactionx11.cpp]: I removed the sleep entirely now without apparent consequences. seems not needed anymore.

File: src/unix/uiactionx11.cpp (hunk 6 above)
Comment (AliKet) [reply src/unix/uiactionx11.cpp:124]: I agree with you that this is a poor design choice and should be change to something better. other than this, **KeyboardEventTestCase** (tests/events/keyboard.cpp) is the only place where the `Default_Delay` is not sufficient and need more additional delay to pass successfully, and IMHO it doesn't worth the trouble to add any new API (at least for now) which won't be needed for the majority of test cases. what do you think ?

Comment (vadz) [reply src/unix/uiactionx11.cpp:124]: The problem is that this env var is also a kind of API, just a hidden and clumsy one, and it doesn't seem preferable to me to just adding the functions in any way.

But looking at this more closely, could we perhaps just always add 10ms for the key events if this is what they need? AFAICS this would make things simpler and what exactly is the drawback of doing it? Just slightly longer test run-time? I think we can live with it...

File: src/unix/uiactionx11.cpp [hunk 11]
```diff
@@ -127,6 +228,10 @@ class wxUIActionSimulatorPlainX11Impl : public wxUIActionSimulatorX11Impl
     explicit wxUIActionSimulatorPlainX11Impl(wxX11Display& display)
         : wxUIActionSimulatorX11Impl(display)
     {
+        wxYield();
+        wxMilliSleep(50);
```
Comment (vadz) [inline src/unix/uiactionx11.cpp:232]: It would be nice to have a symbolic constant for this one too, maybe defined together with the delays used by `wxXSync`, so that they're all in the same place.

File: tests/controls/gridtest.cpp [hunk 12]
```diff
@@ -716,6 +710,9 @@ TEST_CASE_METHOD(GridTestCase, "Grid::SelectUsingEndKey", "[grid]")
 
     sim.KeyDown(WXK_END, wxMOD_CONTROL | wxMOD_SHIFT);
     sim.KeyUp(WXK_END, wxMOD_CONTROL | wxMOD_SHIFT);
+#ifdef __WXGTK20__
+    wxMilliSleep(10);
+#endif
```
Comment (vadz) [inline tests/controls/gridtest.cpp:715]: Looks like this could be removed too if we just used extra 10ms for all emulated key presses?

File: tests/controls/gridtest.cpp (hunk 1 above)
Comment (vadz) [reply tests/controls/gridtest.cpp:1186]: Thanks for the explanation!

Ideal would be to take care of this inside `MouseMove()` implementation itself, i.e. check if the mouse is not in the window there and generate 2 events in this case. And if we can't do it there, we should at least do it in `MouseDragDrop()` where it should definitely be possible and where it would always be needed.

**Review by vadz — APPROVED** [review]
Thanks again, we can already merge this and improve things later, I'm just waiting for the CI builds to pass again on master as they got broken for unrelated reasons in the meanwhile. but if we could get rid of the hack with the env var, it would be really great.

File: src/unix/uiactionx11.cpp (hunk 6 above)
Comment (AliKet) [reply src/unix/uiactionx11.cpp:135]: I'll try to get rid of it then...

File: tests/controls/gridtest.cpp (hunk 1 above)
Comment (AliKet) [reply tests/controls/gridtest.cpp]: Done!

**Comment by vadz** [comment]
I've finally merged this, thanks again!

FWIW I still have some problems when running the tests locally, but just a few (and not dozens if not hundreds as before). I'll look into them later, but, as usual, they can't be reproduced when running them on their own, e.g. `wxDVC::GetItemRect` test fails when run as part of the full test suite, but not when you run just it or even all wxDVC tests. I wish I understood why this happens...

**Comment by AliKet** [comment]
Thanks for merging this. and for the sake of completeness, do you mind to apply this patch too, which i intended to make it part of this PR but failed to do it in time sorry ?

The patch: https://github.com/AliKet/wxWidgets/commit/3741f307c4d3c0ce520317381f78d7a1b3c81ca5.diff

here is the Travis CI run of these changes:
https://travis-ci.org/github/AliKet/wxWidgets/builds/687613687

TIA!
